### PR TITLE
Add ARCHITECTURE.md file

### PR DESCRIPTION
**Pull request by xavdid-stripe**
### Reviewers
r?
cc @stripe/developer-products

 ### Summary

After reading through the code, add an [ARCHITECTURE](https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html) file, outlining the general structure of the codebase and hottest code paths. This was mostly meant as an exercise for myself, but I've found it helpful to save and maintain these documents over time.

Also, I haven't done much work in this codebase, so please double check that what I wrote is accurate and helpful! Also, feel free to suggest other important sections that a new contributor would find helpful when diving into an unfamiliar codebase.

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+`root.go`'s `Execute` function handles actually invoking `Cobra`, via the `rootCmd.ExecuteContext` function. CLI input is parsed and routed to the appropriate command, which runs and/or errors accordingly. If there's an error, `Execute` may print additional help text before exiting.
+
+### Auto-Generated Resources
+
+In addition to the handwritten commands, the CLI also has many auto-generated resources that correspond to base [API resources](https://stripe.com/docs/api/charges). These are generated in `pkg/cmd/resource/resource.go`, mostly via the `NewResourceCmd` function. Those calls come from `pkg/cmd/resources_cmds.go`, which _iself_ is an auto-generated file. It gets a big list of resources from OpenAPI via `spec3.sdk.json` and runs them through `pkg/gen/gen_resource_cmds.go` to write the `.go` files.
+
+### Operation Commands
+
+Under each resource are a set of basic operations, corresponding to CRUD operations for that resource (`retrieve`, `create`, etc). They are built in `pkg/cmd/resource/operation.go` via calls in `resources_cmds.go` to the `NewOperationCmd` method.
```
Comment (tomer-stripe) [inline ARCHITECTURE.md:19]: Might be worth clarifying operation commands are basically for resources only.

**Review by tomer-stripe — COMMENTED** [review]
Thank you for doing this! I'm trying to think through how exactly to do this but we probably want to further tease apart the auto-genned resource commands (and how they work) with non-API commands (like listen, status, etc).

Probably also worth explaining how configs and plugins work, but that's for us and not you/this PR.

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# `stripe-cli` Architecture
+
+The Stripe CLI acts as an ergonomic wrapper around most of the [Stripe API](https://stripe.com/docs/api). It's written in [go](https://go.dev/) and is compiled into a binary for distribution.
+
+## Commands
+
+Its functionality is organized into individual command files, found in `pkg/`. Most files correspond to a command that can be run via the CLI; for example, the `stripe get` is powered by `pkg/get.go`. `pkg/` also includes some test files (`pkg/resources_test.go` has the tests for `pkg/resources.go`, per go convention).
```
[the remark this answers is not in the document]
Comment (xavdid-stripe) [reply ARCHITECTURE.md:7]: ah, bad example then. I was trying to pick a simple-ish command to illustrate "`stripe x` is powered by `pkg/x.go`", but didn't notice the `RunE` param. Seems like `login` is a good example though, I'll swap!

**Comment by xavdid-stripe** [comment]
Ok, I've tweaked the language based on feedback- it's hopefully more clear now!

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+`root.go`'s `Execute` function handles actually invoking `Cobra`, via the `rootCmd.ExecuteContext` function. CLI input is parsed and routed to the appropriate command, which runs and/or errors accordingly. If there's an error, `Execute` may print additional help text before exiting.
+
+### Auto-Generated Resources
+
+In addition to the handwritten commands, the CLI also has many auto-generated resources that correspond to base [API resources](https://stripe.com/docs/api/charges). These commands are registered in `pkg/cmd/resources_cmds.go`, calling the generic `NewNamespaceCmd` function for each resource. These auto-generated commands hold no functionality on their own and mostly rely on [operation commands](#operation-commands) to do anything (see below).
+
+The big list of commands found in `resources_cmds.go` is _iself_ is an auto-generated file. It's built by running `pkg/gen/gen_resource_cmds.go`, which gets a big list of resources from OpenAPI via `api/openapi-spec/spec3.sdk.json` and generates the big command file via the `pkg/gen/resources_cmds.go.tpl` template.
```
Comment (tomer-stripe) [inline ARCHITECTURE.md:17]: We should also add that API commands can be overridden if needed, such as  https://github.com/stripe/stripe-cli/blob/master/pkg/cmd/resource/events.go

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# `stripe-cli` Architecture
+
+The Stripe CLI acts as an ergonomic wrapper around most of the [Stripe API](https://stripe.com/docs/api). It's written in [go](https://go.dev/) and is compiled into a binary for distribution.
+
+## Commands
+
+Its functionality is organized into individual command files, found in `pkg/`. Most files correspond to a command that can be run via the CLI; for example, the `stripe login` is powered by `pkg/login.go`. `pkg/` also includes some test files (e.g. `pkg/open_test.go` has the tests for `pkg/open.go`, per language convention). Some commands have _further_ subcommands, which are defined in their folder under `pkg`. For instance, `pkg/plugin.go` only serves to collect the sub-commands in `pkg/plugins/*.go`.
```
[the remark this answers is not in the document]
Comment (xavdid-stripe) [reply ARCHITECTURE.md:7]: ah, I missed the `cmd` subpath- the paragraph should make more sense now. Sorry about that!

**Comment by xavdid-stripe** [comment]
updated per your comments- thank you!

**Review by vcheung-stripe — APPROVED** [review]

**Pull request merged**